pkg/acl/acl.go

```diff
@@ -0,0 +1,422 @@
+//
+//   acl_create_entry = (int (*)(acl_t *, acl_entry_t *))f;
+//   return acl_create_entry(acl_p, entry_p);
+// }
+//
+// int
+// my_acl_set_tag_type(void *f, acl_entry_t entry_d, acl_tag_t tag_type)
+// {
+//   int (*acl_set_tag_type)(acl_entry_t entry_d, acl_tag_t tag_type);
+//
+//   acl_set_tag_type = (int (*)(acl_entry_t, acl_tag_t))f;
+//   return acl_set_tag_type(entry_d, tag_type);
+// }
+//
+// int
+// my_acl_get_permset(void *f, acl_entry_t entry_d, acl_permset_t *permset_p)
+// {
+//   int (*acl_get_permset)(acl_entry_t entry_d, acl_permset_t *permset_p);
+//
+//   acl_get_permset = (int (*)(acl_entry_t, acl_permset_t *))f;
+//   return acl_get_permset(entry_d, permset_p);
+// }
+//
+// int
+// my_acl_add_perm(void *f, acl_permset_t permset_d, acl_perm_t perm)
+// {
+//   int (*acl_add_perm)(acl_permset_t permset_d, acl_perm_t perm);
+//
+//   acl_add_perm = (int (*)(acl_permset_t, acl_perm_t))f;
+//   return acl_add_perm(permset_d, perm);
+// }
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"unsafe"
+)
+
+const (
+	otherExec  = 1 << iota
+	otherWrite = 1 << iota
+	otherRead  = 1 << iota
+	groupExec  = 1 << iota
+	groupWrite = 1 << iota
+	groupRead  = 1 << iota
+	userExec   = 1 << iota
+	userWrite  = 1 << iota
+	userRead   = 1 << iota
+
+	TagUserObj  Tag = C.ACL_USER_OBJ
+	TagUser     Tag = C.ACL_USER
+	TagGroupObj Tag = C.ACL_GROUP_OBJ
+	TagGroup    Tag = C.ACL_GROUP
+	TagMask     Tag = C.ACL_MASK
+	TagOther    Tag = C.ACL_OTHER
+
+	PermRead    Perm = C.ACL_READ
+	PermWrite   Perm = C.ACL_WRITE
+	PermExecute Perm = C.ACL_EXECUTE
+)
+
+type ACL struct {
+	lib *libHandle
+	a   C.acl_t
+}
+
+// Entry is an entry in an ACL.
+type Entry struct {
+	a *ACL
+	e C.acl_entry_t
+}
+
+// Permset is a set of permissions.
+type Permset struct {
+	a *ACL
+	p C.acl_permset_t
+}
+
+// Perm represents a permission in the e_perm ACL field
+type Perm int
+
+// Tag represents an ACL e_tag entry
+type Tag int
+
+var ErrSoNotFound = errors.New("unable to open a handle to libacl")
+
+type libHandle struct {
+	handle  unsafe.Pointer
+	libname string
+}
+
+func getHandle() (*libHandle, error) {
+	for _, name := range []string{
+		"libacl.so",
+		"libacl.so.1",
+	} {
+		libname := C.CString(name)
+		defer C.free(unsafe.Pointer(libname))
+		handle := C.dlopen(libname, C.RTLD_LAZY)
+		if handle != nil {
+			h := &libHandle{
+				handle:  handle,
+				libname: name,
+			}
+			return h, nil
+		}
+	}
+	return nil, ErrSoNotFound
+}
+
+func (h *libHandle) close() error {
+	if r := C.dlclose(h.handle); r != 0 {
+		return fmt.Errorf("error closing %v: %d", h.libname, r)
+	}
+	return nil
+}
+
+// InitACL dlopens libacl and returns an ACL object if successful.
+func InitACL() (*ACL, error) {
+	h, err := getHandle()
+	if err != nil {
+		return nil, err
+	}
+
+	return &ACL{lib: h}, nil
+}
+
+// ParseACL parses a string representation of an ACL.
+func (a *ACL) ParseACL(acl string) (err error) {
+	sym := C.CString("acl_from_text")
```

Comment:
How about having a function taking a symbol name and a handle and returning the function pointer and an error? That would cut down code duplication slightly.